src/mysqlsh_helpers.py

```diff
@@ -0,0 +1,176 @@
+#!/usr/bin/env python3
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Helper class to manage the MySQL InnoDB cluster lifecycle with MySQL Shell."""
+
+import logging
+
+from ops.model import Container
+from ops.pebble import ExecError
+from tenacity import retry, stop_after_delay, wait_fixed
+
+logger = logging.getLogger(__name__)
+
+
+class MySQLInstanceConfigureError(Exception):
+    """Exception raised when there is an issue configuring a MySQL instance."""
+
+    pass
+
+
+class MySQLConfigureMySQLUsersError(Exception):
+    """Exception raised when creating a user fails."""
+
+    pass
+
+
+class MySQL:
+    """Class to encapsulate all operations related to the MySQL instance and cluster.
+
+    This class handles the configuration of MySQL instances, and also the
+    creation and configuration of MySQL InnoDB clusters via Group Replication.
+    """
+
+    def __init__(
+        self,
+        root_password: str,
+        cluster_admin_user: str,
+        cluster_admin_password: str,
+        instance_address: str,
+    ):
+        """Initialize the MySQL class.
+
+        Args:
+            root_password: password for the 'root' user
+            cluster_admin_user: user name for the cluster admin user
+            cluster_admin_password: password for the cluster admin user
+            instance_address: address of the targeted instance
+        """
+        self.root_password = root_password
+        self.cluster_admin_user = cluster_admin_user
+        self.cluster_admin_password = cluster_admin_password
+        self.instance_address = instance_address
+
+    def mysqlsh_bin(self, container: Container) -> str:
+        """Determine binary path for MySQL Shell.
+
+        Args:
+            container: Container to run the script in
+
+        Returns:
+            Path to binary mysqlsh
+        """
+        # Allow for various versions of the mysql-shell snap
+        # When we get the alias use /snap/bin/mysqlsh
+        _paths = ("/snap/bin/mysqlsh", "/snap/bin/mysql-shell.mysqlsh")
+
+        for path in _paths:
+            if container.exists(path):
+                return path
+        # Default to the upstream repo mysql-shell
+        return "/usr/bin/mysqlsh"
+
+    def configure_mysql_users(self) -> None:
+        """Configure the MySQL users for the instance.
+
+        Creates base `root@%` and `clusteradmin@instance_address` user with the
+        appropriate privileges, and reconfigure `root@localhost` user password.
+        Raises MySQLConfigureMySQLUsersError if the user creation fails.
+        """
+        _script = (
+            "SET @@SESSION.SQL_LOG_BIN=0;",
+            f"CREATE USER '{self.cluster_admin_user}'@'{self.instance_address}' IDENTIFIED BY '{self.cluster_admin_password}';",
```

Comment:
two comments:

1. we should create the `serverconfig@%` user manually, and then use `serverconfig@%` user to call `dba.configure_instance`. it is while configuring the instance that we create the `clusteradmin@%` user by passing in `clusterAdmin` and `clusterAdminPassword` parameters to `configure_instance`
2. we should avoid using `mysql_native_password` -> to do this, we may need to break the commands into 2 groups of commands (still to be verified by Mykola in the mysql-operator repository)

Reply:
I guess I was waiting for the merge in machine land, but I can use the branch code anyway right.